File: 2021/Rust/src/day02.rs
use std::time::Instant;

use eyre::Result;

struct Instruction {
    dir: Direction,
    amount: u64,
}

#[derive(Debug, Clone, Copy)]
enum Direction {
    Up,
    Down,
    Forward,
}

fn parse(input: &str) -> Vec<Instruction> {
    input
        .lines()
        .map(|line| {
            let mut i = line.split(' ');
            let dir = match i.next() {
                Some("up") => Direction::Up,
                Some("down") => Direction::Down,
                Some("forward") => Direction::Forward,
                _ => unreachable!("invalid input"),
            };
            let amount = i
                .next()
                .and_then(|x| x.parse().ok())
                .expect("invalid input");

            debug_assert!(i.next().is_none());

            Instruction { dir, amount }
        })
        .collect()
}

fn part1(input: &[Instruction]) -> u64 {
    let mut depth = 0;
    let mut horiz = 0;

    for Instruction { dir, amount } in input {
        match dir {
            Direction::Forward => horiz += amount,
            Direction::Up => depth -= amount,
            Direction::Down => depth += amount,
        }
    }

    depth * horiz
}

fn part2(input: &[Instruction]) -> u64 {
    let mut depth = 0;
    let mut horiz = 0;
    let mut aim = 0;

    for Instruction { dir, amount } in input {
        match dir {
            Direction::Forward => {
                horiz += amount;
                depth += amount * aim;
            }
            Direction::Up => aim -= amount,
            Direction::Down => aim += amount,
        }
    }

    depth * horiz
}

pub fn run(input: String) -> Result<()> {
    let parsed = parse(input.as_str());
    let start = Instant::now();
    let part1 = part1(&parsed);
    let p1_time = Instant::now() - start;
    println!(
        "The solution to 2021 day 02 (part 1) is {}. Took {:?}",
        part1, p1_time
    );
    let start = Instant::now();
    let part2 = part2(&parsed);
    let p2_time = Instant::now() - start;
    println!(
        "The solution to 2021 day 02 (part 2) is {}. Took {:?}",
        part2, p2_time
    );

    Ok(())
}

File: 2021/Haskell/src/Day02.hs
{-# LANGUAGE OverloadedStrings #-}

module Day02
  ( day02,
  )
where

import AoC
import Common
import Data.Functor (($>))
import Data.Text (Text)
import qualified Data.Text as Text
import Text.Megaparsec
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L

type Parsed = [Instruction]

data Direction = Forward | Down | Up deriving (Show)

data Instruction = Instruction
  { direction :: Direction,
    amount :: Int
  }
  deriving (Show)

pInstruction :: Parser Instruction
pInstruction = do
  direction <- pDirection
  _ <- char ' '
  amount <- L.decimal
  return Instruction {..}

pDirection :: Parser Direction
pDirection =
  string "up" $> Up
    <|> string "down" $> Down
    <|> string "forward" $> Forward

moveAbs :: [Instruction] -> (Int, Int)
moveAbs instrs = go instrs 0 0
  where
    go (a : xs) depth horiz = case a of
      Instruction Forward x -> go xs depth (horiz + x)
      Instruction Down x -> go xs (depth + x) horiz
      Instruction Up x -> go xs (depth - x) horiz
    go [] depth horiz = (depth, horiz)

moveAim :: [Instruction] -> (Int, Int)
moveAim instrs = go instrs 0 0 0
  where
    go (a : xs) aim depth horiz = case a of
      Instruction Forward x -> go xs aim (depth + (aim * x)) (horiz + x)
      Instruction Down x -> go xs (aim + x) depth horiz
      Instruction Up x -> go xs (aim - x) depth horiz
    go [] _ depth horiz = (depth, horiz)

day02 :: Runner Parsed Int
day02 =
  let year = 2021
      day = 2
      parser :: Text -> Maybe Parsed
      parser input = unwrapParser $ traverse (parse pInstruction "input") (Text.lines input)
      part1 = return . uncurry (*) . moveAbs
      part2 = return . uncurry (*) . moveAim
   in Runner {..}
